### DISCUSSION TITLE
Realtime: listening to row level changes doesn't work

### DISCUSSION THREAD

**Discussion by orange472** [supabase/supabase#8005] [Questions] [answered]
Hello!

I am trying to set up a realtime listener for one of my tables called "Room" and was hoping to get some guidance. So far, I figured out how to listen to the entire table but I'm struggling with specific rows. Here is my code below:

```
React.useEffect(() => {
    if (props.roomId == undefined) return; // I made sure the roomId here exists in a row in the table

    const subscription = supabase
      .from(`Room:roomId=eq.${props.roomId}`) // .from("Room") works but listens to the entire table
      .on("UPDATE", (payload) => {
        console.log("Room updated");
      })
      .subscribe();

    const channel = supabase.channel(
      `realtime:public:Room:roomId=eq.${props.roomId}`, // similarly, .channel("realtime:public:Room") works
      { selfBroadcast: true }
    );
    channel.subscribe().receive("ok", () => {
      console.log("Room channel created");
    });

    // both subscription and channel have a state of "joining" when I console.log them; I set an interval to see if the state would change, but it doesn't

    // clean up, BTW: React.StrictMode is disabled
    return () => {
      subscription.unsubscribe();
      supabase.removeSubscription(subscription);

      channel.unsubscribe();
      supabase.removeChannel(channel);
    };
  }, []);
```

Any help would be appreciated!

**Comment by GaryAustin1** [comment, marked as answer]
Your general format looks correct.  Caps seem to work fine for both table and column (which was my first guess, but I tested it).
Verify your column name in the table is roomId single cap I?
Is roomId your primary key?  Do you have a primary key?
Does it work if you just run the first supabase-js subscription and not also do the 2nd direct  realtime subscription?

**Reply by orange472** [reply]
Hi, thank you so much for responding!

I made sure that the column name in the table is written as roomId (with one capital I) and that roomId is the primary key. 

I tried writing the channel names without "eq." (for example, `Room:id=${props.roomId}` instead of `Room:id=eq.${props.roomId}`). When I refreshed the page, I got the expected console message: "Room channel created" for the second realtime subscription, but not the first. I logged both subscriptions and got a state of "closed" for the first and "joined" for the second. 

```
// first subscription
const subscription = supabase
      .from(`Room:roomId=${props.roomId}`)
      .on("UPDATE", (payload) => {
        console.log("Update:", payload);
      })
      .subscribe();
// console.log(subscription) => state: "closed"

// second subscription
const roomChannel = supabase.channel(
    `realtime:public:Room:roomId=${props.roomId}`,
    { selfBroadcast: true }
);

roomChannel.subscribe().receive("ok", () => {
    console.log("Added broadcaster listener");
    // console.log(roomChannel) -> state: "joined"
});
```
Side question: I want the second subscription for broadcasting. However, roomChannel is an instance of RealtimeSubscription, not RealtimeChannel, so I don't have access to broadcasting.
```
React.useEffect(() => {
    if (roomChannel != undefined) {
      console.log(roomChannel.constructor.name); // "RealtimeSubscription"
      console.log(roomChannel instanceof RealtimeChannel); // false
      console.log(typeof roomChannel.send); // undefined
    }
  }, [roomChannel]);
```
The description of the supabase.channel method is "Creates a channel with Broadcast and Presence. Activated when vsndate query param is present in the WebSocket URL." What is a vsndate query param and how can I provide it?

**Reply by GaryAustin1** [reply]
I don't see how roomId=  works without  eq.  but have never tried it.
One thing to note is if you subscribe to the exact same thing (table or table with filter) the first connection will be closed.  SB (at least for sure supabase-js only allows one channel for same thing.

**Reply by orange472** [reply]
I got rid of the second subscription and now I'm getting "joined" for the first subscription, but I think "joined" is a little misleading; it turns out that I can type anything after the table name (e.g. Room:bananas) and the subscription will have a state of "joined" but will not respond to db changes.

I get "joining" for `Room:roomId=eq.${roomId}` though, which is kind of a good sign, so I added a callback to .subscribe() in order to log errors and I get "error occurred when joining <table_name> with user token". 

Also if I just do `supabase.from("Room").on("UPDATE", (payload) => { console.log(payload }).subscribe()`, the payload object has this error: "Error 401: Unauthorized", even though my RLS literally allows anyone to do anything and I'm signed in.

Any suggestions?

**Reply by GaryAustin1** [reply]
I think the 401 error is new to help notice when Postgres grants on the table or schemas is not set to allow realtime access.  Was added mainly because of Prisma and custom schema needed grants set up.    

Anything special about your table or schema setup?  Did you change the access grants?  

If not you should probably contact support to look.  

Edit:  The actual definition of the error seems broader than just grants on the table/schema.
```
Error 401: Unauthorized

If a WAL record is passed through realtime.apply_rls and the subscription's clams_role does not have permission to select the primary key columns in that table, an Unauthorized error is returned with no WAL data.
```

**Reply by LewTrn** [reply]
> I got rid of the second subscription and now I'm getting "joined" for the first subscription, but I think "joined" is a little misleading; it turns out that I can type anything after the table name (e.g. Room:bananas) and the subscription will have a state of "joined" but will not respond to db changes.
> 
> I get "joining" for `Room:roomId=eq.${roomId}` though, which is kind of a good sign, so I added a callback to .subscribe() in order to log errors and I get "error occurred when joining <table_name> with user token".
> 
> Also if I just do `supabase.from("Room").on("UPDATE", (payload) => { console.log(payload }).subscribe()`, the payload object has this error: "Error 401: Unauthorized", even though my RLS literally allows anyone to do anything and I'm signed in.
> 
> Any suggestions?

I am also seeing this exact behaviour in a recent project. I am able to listen to table level changes (although I don't see the whole payload as I get the same 401 Unauthorized error). However, if I try to subscribe to a specific row, then modify it, I see no update. 😕 

There's nothing particularly special about my tables or schemas, other than the fact I used a Prisma setup.

**Reply by GaryAustin1** [reply]
@LewTrn is it possibly this:  https://supabase.com/docs/guides/integrations/prisma#troubleshooting
Prisma messes with the  grants on the public schema.

**Reply by orange472** [reply]
I also used Prisma and your link helped me out greatly. Thank you!
